### DISCUSSION TITLE
Journey: Headers not defined & V3 Support & Custom Query Options

### DISCUSSION THREAD

**Discussion by szarcensky** [odata2ts/odata2ts#217] [Q&A] [answered]
After generating code using mode all, any attempt to build the code returns the following error:

`Class constructor QueryObject cannot be invoked without 'new'`

The error comes from every instance of this line of code, found in the Q-Objects:

`var _this = _super !== null && _super.apply(this, arguments) || this;`

I have target set to es2021 in tsconfig.json but the problem remains.

**Comment by texttechne** [comment]
Hi @szarcensky,

I definitely need more information. Can you produce a minimal example?
At least, I would need to see your `odata2ts.config.ts`, your metadata file and how you instantiate  and use stuff...

On your part, you could also test the following:
- Does the same error occur if you use one of the [reference services](https://www.odata.org/odata-services/)?

**Comment by szarcensky** [comment]
So I tried the reference service, and that error doesn't show up. I then switched back to the odata service I was originally using, and I cannot get this error any longer. It's not creating that code any longer. I'm instead getting a bunch of new errors I cannot solve. 

It may be easier to just explain the very simple thing I'm trying to do, that I've spent weeks now failing to do, and maybe you can just tell me what I'm doing wrong. I literally just want any piece of data to come back from the odata service, to prove it's working on a basic level. 

I'm on Node 20.5.1 because I grabbed whatever was current. 

My odata2ts.config.ts is just this

```
import { ConfigFileOptions, Modes, EmitModes } from '@odata2ts/odata2ts';
const config: ConfigFileOptions = {
  mode: Modes.all,
  emitMode: EmitModes.ts,
  services: {
    rhymba: {
      sourceUrl: 'https://search.mcnemanager.com/dev/content.odata/',
      source: 'src/resource/rhymba.xml',
      output: 'src/build/rhymba',
      prettier: true,
    },
  },
};
export default config;
```

My goal is to make an SDK for the odata API, so I googled for boilerplate NPM setups and got this one: https://github.com/tomchen/example-typescript-package

I tried running it as a test using Jest, and I tried building js files and linking them to install in a separate local new Angular project. I've tried using EmitModes.ts and the default (I'll ultimately want both, as I need a TS version and a JS version).

Right now, the TS files are giving me this error, which makes zero sense to me:
`ReferenceError: Headers is not defined`
on this line
`this.httpClient = new FetchClient();`

If I generate with JS files instead, I get this error:
`SyntaxError: Cannot use import statement outside a module`
from simply importing the main service.

**Comment by texttechne** [comment]
Hi @szarcensky,

doesn't sound like a nice journey. 

First off, you usually want to use Node.js in its `LTS` version which stands for "Long Term Support". The other version is bleeding edge...
So you should use 18.x currently. And as best practice:
* remove your current node installation
* install NVM (Node Version Manager) first
  * Windows: [nvm-for-windows](https://github.com/coreybutler/nvm-windows/releases)
  * *Unix: [nvm](https://github.com/nvm-sh/nvm)
* install and use the node versions you want to use
  * could be as easy as `nvm install lts` and `nvm use lts` (however read their docs..)

That's my first tip.

**Comment by texttechne** [comment, marked as answer]
Hi @szarcensky,

> ReferenceError: Headers is not defined
This kind of error can occur in the combination Jest (<29.x) + FetchClient. As Jest runs as a node process without any browser, it's kind of a special situation when it comes down to the usage of `Fetch`. 

The Fetch API was originally only available in browsers, but since Node.js v17.x it has been ported and activated by default. So the [Headers](https://developer.mozilla.org/en-US/docs/Web/API/Headers) - used by our FetchClient- should be known to JS run in Node. 

However, Jest used to handle global stuff in a specific manner, which brings you this error. Updating `Jest` & `ts-jest` to `29.x` resolves this issue as they've changed jest to expose global stuff in a different manner.

Here is my test repo, which builds on your given use case: [Test Repo](https://github.com/texttechne/test-ts-setup). Take a look at my commits, the `jest.config.js` also needed to be updated.

Final Result: Your service is currently not working 😄

**Comment by texttechne** [comment]
Regarding to what you're trying to achieve right now:
>  I literally just want any piece of data to come back from the odata service, to prove it's working on a basic level.

To only do that, i.e. without any automated tests, with the sole focus on developing the OData server, I would recommend to use a proper REST client. 
* VsCode: the plugin is called `REST Client` 
* otherwise something like Postman
* maybe you're into Curl 😉?
* ...

See my [http tests](https://github.com/odata2ts/odata2ts/tree/main/test/v2) as an example for the VsCode REST client scripts.

**Comment by szarcensky** [comment]
@texttechne thanks for all this info. Especially the rest repo, being able to compare to a working demo with my use case is going to make my life so much easier.

I already use NVM, didn't realize I should be using LTS. 

The Jest info is super useful, I think that'll help a lot. 

The odata service is already in production, though we're fixing some things I didn't realize were done incorrectly until I had to consume the service myself. 

I'll try all the things you've mentioned here and let you know how it goes.

Also, is there a way to pass a query string to the Fetch client?

**Comment by texttechne** [comment]
>Also, is there a way to pass a query string to the Fetch client?

Globally, i.e. for all requests, via the first constructor param:
`new FetchClient({params: {myParam: myValue}})`

Per request:
`.query( undefined, {params: {myParam: myValue}})`

**Comment by szarcensky** [comment]
So I just realized your readme says odata v2 and v4, and my service is v3. The class structure is expecting the response in the format `response.data.d.results` but what's coming back from the odata service is in the format `response.data.value`. Is there something I can do to work around that?

**Reply by texttechne** [reply]
Hi @szarcensky,

V3 is currently not supported by odata2ts. Well, I've never tested that at least. 

When developing I found out that `ComplexType` and `Enum` were invented with V3. This SHOULD be supported.

However, the response types are a different matter. Could you please verify that they are aligned with the [V4 reponse types from the core package](https://github.com/odata2ts/odata2ts/blob/main/packages/odata-core/src/ResponseModelV4.ts)?

If so I should be able to support V3 in this regard...

**Reply by szarcensky** [reply]
Both are using `value` to hold the data, but V3 uses `odata.metadata` instead of `@odata.context`, and `odata.count` instead of `@odata.count`. Paging also using `odata.nextLink` instead of `@odata.nextLink`, so mostly it's not using the `@` prefix.

[Northwind V3](https://services.odata.org/V3/Northwind/Northwind.svc/Customers?$format=json)
[Northwind V4](https://services.odata.org/V4/Northwind/Northwind.svc/Customers?$format=json)

**Reply by szarcensky** [reply]
Did you need any additional info from me about V3 vs V4 response types? Does supporting V3 seem viable?

**Reply by texttechne** [reply]
Yes V3 support seems viable. Would be cool if you could create an own issue for that; now that we've sorted out the other things.

I think we can close this discussion then...

**Comment by szarcensky** [comment]
One more question that might be trickier. The odata service has a couple non standard parameters on top of the base odata stuff. They don't start with a dollar sign, and show up as parameters in the function import. 

```
        <FunctionImport
          Name="Media"
          ReturnType="Collection(Rhymba.MediaEntity)"
          EntitySet="MediaModel"
          m:HttpMethod="GET"
        >
          <Parameter Name="id_cdl" Type="Edm.String" />
          <Parameter Name="keyword" Type="Edm.String" />
        </FunctionImport>
```

It looks like I can pass them as `params` when instantiating `Media`, but I don't see how I would use them with `MediaModel`, or where I'd really like them to work, as an operation in the query builder. 

Is there a way to do that with the current setup? Or if that's something I could extend in a separate file? I see that I can add it to `ODataQueryBuilderModel.d.ts`, but it's not recognized as a function when I try to run it, which makes sense.

**Reply by texttechne** [reply]
I'm a bit confused:
> The odata service has a couple non standard parameters on top of the base odata stuff.

Do you mean that the `MediaModel` has extra properties which are not advertised by the metadata?

> I don't see how I would use them with MediaModel, or where I'd really like them to work, as an operation in the query builder.

If you really mean non-advertised props, there's not much odata2ts can do for you. You can still use the query builder but you would need to write the filter operation yourself: see [Custom Filter Expressions](https://odata2ts.github.io/docs/query-builder/filtering#custom-filter-expressions)

**Reply by texttechne** [reply]
Regarding the `FunctionImport`. odata2ts generates an appropriate function into the generated service. By calling the generated function you call the corresponding function of your OData service, executing a request....

**Reply by szarcensky** [reply]
`id_cdl` and `keyword` are the two parameters I was talking about. They show up in the function import, so they are in the metadata. I don't see them generated as functions, but they're not used as stand alone, but as query parameters, the same way you'd call `top` or `filter`. I'd hoped to be able to do something like this:

```
      const response: any = await rhymbaService
        .MediaModel()
        .query((builder, qMedia) => {
          return builder
            .select('id')
            .select('title')
            .select('geos')
            .filter(qMedia.artist_name.eq('Crystal Valenzuela'))
            .keyword('spin')
            .expand('geos')
            .top(4)
        })
```

But they're not being generated in a way that allows that. I only see them in the Q Model and Params:

QRhymba.ts:
```
export class QMedia extends QFunction<MediaParams> {
  private readonly params = [
    new QStringParam('id_cdl'),
    new QStringParam('keyword'),
  ]

```
RhymbaModel.ts:
```
export interface MediaParams {
  id_cdl?: string | null
  keyword?: string | null
}

```

**Reply by texttechne** [reply]
I think you've got something confused here, but I cannot pinpoint what it is.

First off, `Media` is a `Function` (that's what the metadata is telling, because it is a FunctionImport). This function has these two parameters:
- id_cdl
- keyword
And this is what you see in QRhymba: the representation of that function called `QMedia`. Additionally, you see in RhymbaModel a model representing the parameters of that function called `MediaParams`. All of which is required to call the `media` function from the generated main service...

Otherwise you have an `EntityType` called `MediaModel` in your metadata. And this is the stuff with which you perform queries. In this `MediaModel` there do not exist any properties called `id_cdl` or `keyword`. They are simply not existing. However, odata would allow you to use properties which have not been advertised by the metadata... I still don't know if that is the case here...

Both things - the `Media` function and the `MediaModel` EntityType - are completely separate topics.

And now you are talking about [custom query options](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_QueryOptionExtensibility), when you mention this example:
```ts
      // don't type the response as any, by the way, since odata2ts will generate the appropriate type here
      const response = await rhymbaService
        .MediaModel()
        .query((builder, qMedia) => builder.keyword('spin'))
```
I don't understand the meaning of this custom "keyword" operation. And is this really the feature you're after?

To move forward, please state the feature you are talking about. Ideally by citing the OData V2 / V3 spec...
Give me an example of the URL your OData server would be able to process regarding this feature.
The last example (custom query option) would translate to:
`/MediaModel?keyword='spin'`
Does this work for your OData service?

**Reply by szarcensky** [reply]
Yes.
`/Media()/?keyword='spin'&$filter=artist_name eq 'Crystal Valenzuela'&$skip=0&$top=4&$select=id,title,artist_name,album_name&access_token=`
returns
```
{
  "odata.metadata": "https://search.mcnemanager.com/dev/content.odata/$metadata#MediaModel&$select=id,title,artist_name,album_name",
  "value": [
    {
      "id": 1205202071,
      "title": "Spin Head",
      "artist_name": "Crystal Valenzuela",
      "album_name": "A Happy Life"
    },
    {
      "id": 1205131408,
      "title": "Spin Head",
      "artist_name": "Crystal Valenzuela",
      "album_name": "A Happy Life"
    }
  ]
}
```
where the results all contain the keyword 'spin'. There would be 4 results instead of 2 for this query without the keyword applied. I believe the idea was to create something similar to V4's search filter. Is the service advertising this incorrectly in the metadata? I may be able to have that changed, if that would be more correct, and prevent the need for a workaround via Custom Query Options on the front end.

**Reply by texttechne** [reply]
Hi,
ok, then we are actually talking about custom query options, which are simply query parameters.
You can add those on a per request basis:
```
rhymbaService
  .mediaModel()
  .query(
    (builder, qMedia) => {
      return builder; // add all your typical filters here
    }, 
    {  params: { keyword: "'spin'"} }  // note the single quotes within the double quotes
  )
```

But be aware that you have to take care of constructing the correct query param value. In the example you have to add single quotes because `keyword` is of type `Edm.String`.

So you would need to know the basic rules for V2 (assuming that they still hold true for V3). They work along these lines:
* string values are surrounded by single quotes: `'spin'`
* some numeric types are suffixed with a letter, e.g for double `12.33d`
* "advanced" types use a prefix and single quotes, e.g. `datetime'2023-10-01'`
* all other types are written literally, e.g. `true`
But check the specification.

odata2ts cannot know about this custom query option as it is NOT advertised in the metadata. So I don't see a way how odata2ts can help here... Hence you have to roll up your sleeves and write this manually. 

If there is a way to include custom query option definitions into the metadata, then please point to the appropriate section in the V2 / V3 / V4- specification as reference.
